src/main/java/cd/go/contrib/elasticagent/KubernetesClientFactory.java

```diff
@@ -31,31 +34,34 @@ public static KubernetesClientFactory instance() {
         return KUBERNETES_CLIENT_FACTORY;
     }
 
-    private static KubernetesClient createClient(PluginSettings pluginSettings) throws Exception {
-        ConfigBuilder configBuilder = new ConfigBuilder().withMasterUrl(pluginSettings.getKubernetesClusterUrl());
-        if (StringUtils.isNotBlank(pluginSettings.getKubernetesClusterUsername())) {
-            configBuilder.withUsername(pluginSettings.getKubernetesClusterUsername());
+    public synchronized KubernetesClient client(PluginSettings pluginSettings) {
+        if (pluginSettings.equals(this.pluginSettings) && this.client != null) {
+            LOG.debug("Using previously created client.");
+            return this.client;
         }
 
-        if (StringUtils.isNotBlank(pluginSettings.getKubernetesClusterPassword())) {
-            configBuilder.withPassword(pluginSettings.getKubernetesClusterPassword());
-        }
+        LOG.debug(format("Creating a new client because {0}.", (client == null) ? "client is null" : "plugin setting is changed"));
+        this.pluginSettings = pluginSettings;
+        this.client = createClientFor(pluginSettings);
+        LOG.debug(format("New client is created using authentication strategy {0}.", pluginSettings.getAuthenticationStrategy()));
+        return this.client;
+    }
 
-        if (StringUtils.isNotBlank(pluginSettings.getKubernetesClusterCACert())) {
-            configBuilder.withCaCertData(pluginSettings.getKubernetesClusterCACert());
-        }
+    private KubernetesClient createClientFor(PluginSettings pluginSettings) {
+        LOG.debug(format("Creating config using authentication strategy {0}.", pluginSettings.getAuthenticationStrategy().name()));
+        final ConfigBuilder configBuilder = new ConfigBuilder();
 
-        Config build = configBuilder.build();
-        return new DefaultKubernetesClient(build);
-    }
+        if (pluginSettings.getAuthenticationStrategy() == OAUTH_TOKEN) {
+            configBuilder.withOauthToken(pluginSettings.getOauthToken());
```

Comment:
Shouldn't the master url be set here too?

Reply:
No we don't need it.

Reply:
That means that the cluster url field need not be mandatory.

Reply:
Yes it not mandatory when authentication strategy is set to `oauth_token`. But mandatory when authentication strategy is set to `cluster_certs`

Reply:
We saw that cluster ip was mandatory. Please change the UI and the field to make it mandatory.